src/text/raw_text_writer.rs

```diff
@@ -9,15 +9,134 @@ use crate::raw_symbol_token_ref::{AsRawSymbolTokenRef, RawSymbolTokenRef};
+    /// ```ignore
+    /// {
+    ///     foo: 1,
+    ///     bar: 2,
+    ///     baz: 3
+    /// }
+    /// [
+    ///     1,
+    ///     2,
+    ///     3
+    /// ]
+    /// true
+    /// "hello"
+    /// ```
+    pub fn pretty() -> RawTextWriterBuilder {
+        RawTextWriterBuilder {
+            // Each value appears on its own line
+            space_between_values: String::from("\n"),
+            // Values get four spaces of indentation per level of depth
+            indentation: String::from("    "),
+            // Field names and values are separated by a single space
+            space_after_field_name: String::from(" "),
+            // The first value in a container appears on a line by itself
+            space_after_container_start: String::from("\n"),
+        }
+    }
+
+    pub fn with_space_between_values<S: Into<String>>(
+        mut self,
+        space_between_values: S,
+    ) -> RawTextWriterBuilder {
+        self.space_between_values = space_between_values.into();
+        self
+    }
+
+    pub fn with_indentation<S: Into<String>>(mut self, indentation: S) -> RawTextWriterBuilder {
+        self.indentation = indentation.into();
+        self
+    }
+
+    pub fn with_space_after_field_name<S: Into<String>>(
+        mut self,
+        space_after_field_name: S,
+    ) -> RawTextWriterBuilder {
+        self.space_after_field_name = space_after_field_name.into();
+        self
+    }
+
+    pub fn with_space_after_container_start<S: Into<String>>(
+        mut self,
+        space_after_container_start: S,
+    ) -> RawTextWriterBuilder {
+        self.space_after_field_name = space_after_container_start.into();
+        self
+    }
+
+    /// Constructs a new instance of [RawTextWriter] that writes values to the provided io::Write
+    /// implementation.
+    pub fn build<W: Write>(self, sink: W) -> IonResult<RawTextWriter<W>> {
+        let raw_text_writer = RawTextWriter {
+            output: BufWriter::new(sink),
+            annotations: Vec::new(),
+            field_name: None,
+            containers: vec![EncodingLevel::default()],
+            string_escape_codes: string_escape_code_init(),
+            // TODO: We should consider putting these in a single struct (`WhitespaceConfig`?)
+            //       and `Box`ing it. Each `String` is 24 bytes, making the writer much bigger.
+            space_between_values: self.space_between_values,
+            indentation: self.indentation,
+            space_after_field_name: self.space_after_field_name,
+            space_after_container_start: self.space_after_container_start,
+        };
+        // This method cannot currently fail. It returns an IonResult<_> to be consistent with the
+        // other builder APIs and to allow for fallible setup operations in the future.
+        Ok(raw_text_writer)
+    }
+}
+
+#[derive(Debug, PartialEq, Default)]
+struct EncodingLevel {
+    container_type: ContainerType,
+    child_count: usize,
+}
```

Comment:
🗺️ The writer needs to track how many values have been written in each container so it knows whether to write a delimiter (`,`) before the next one.